Задание 1			
Скачать данные по ссылке
 								
https://www.kaggle.com/datasets/ionaskel/laptop-prices				
1. 1 Создать новый признак Cpu_Company, который будет содержать только название фирмы, которая произвела CPU (Intel, AMD и т.д)	
1. 2 Создать новый признак Memory_Amount, который будет содержать только количество Gb памяти без указания типа носителя (128, 1024, 256 и т.д.)					
1. 3 Создать новый признак Memory_Type, который будет содержать только тип носителя (HDD, SDD и т.д.)					
1. 4 Удалите признаки Memory и ScreenResolution 
Задание 2.		
2. 1 Создайте признак SSD, который изначально равен 0		
2. 2 Поставьте в признаке SSD 1, если ноутбук действительно с типом носителя SSD						
2. 3 Уберите в признаке Weight значения 'kg' и поменяйте его тип данных на вещественный 					
Задание 3.		
Создайте датафрейм с клиентами:					
clients = pd.DataFrame({
'client_id': [45, 32, 67, 33, 43], 'laptop_id': [506, 398, 710, 120, 1999]		
})				
laptop_id - это индексы датафрейма с ноутбуками			
3. 1 Присоедините к таблице clients данные по ноутбукам через метод join						
3. 2 Присоедините к таблице clients данные по ноутбукам через метод merge					
Это нужно, чтобы понимать, какие ноутбуки покупались клиентам
Задание 4.					
Составьте несколько сводных таблиц				
4. 1 Найдите среднюю стоимость ноутбуков в зависимости от компании производителя		
Отсортируйте от меньшей стоимости к большей		
4. 2 Найдите минимальную, среднюю и максимальную стоимости ноутбуков в зависимости от производителя процессора	
4. 3 Постройте таблицу с подсчетом количества ноутбуков в данных в зависимости от производителя CPU и ОЗУ					
4. 4 Постройте таблицу с подсчетом средней стоимости ноутбуков в данных в зависимости от операционной системы и GB памяти 
Задание 5*.		
Ответьте на несколько вопросов, используя методы группировки	
5. 1 Ноутбуков каких компаний и с каким процессором больше?	
5. 2 С каким типом памяти и с каким объемом памяти больше ноутбуков? 


In [45]:
import pandas as pd

df = pd.read_csv('./laptops.csv',encoding='ISO-8859-1',index_col=[0])
df.head(5)


Company      Product   TypeName  Inches                    ScreenResolution  \
1   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   
2   Apple  Macbook Air  Ultrabook    13.3                            1440x900   
3      HP       250 G6   Notebook    15.6                   Full HD 1920x1080   
4   Apple  MacBook Pro  Ultrabook    15.4  IPS Panel Retina Display 2880x1800   
5   Apple  MacBook Pro  Ultrabook    13.3  IPS Panel Retina Display 2560x1600   

                          Cpu   Ram               Memory  \
1        Intel Core i5 2.3GHz   8GB            128GB SSD   
2        Intel Core i5 1.8GHz   8GB  128GB Flash Storage   
3  Intel Core i5 7200U 2.5GHz   8GB            256GB SSD   
4        Intel Core i7 2.7GHz  16GB            512GB SSD   
5        Intel Core i5 3.1GHz   8GB            256GB SSD   

                            Gpu  OpSys  Weight  Price_euros  
1  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69  
2        Intel HD Graphics 6000  macOS  1.34kg       898.94  
3         Intel HD Graphics 620  No OS  1.86kg       575.00  
4            AMD Radeon Pro 455  macOS  1.83kg      2537.45  
5  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60

In [12]:
df['CPU_Company'] = df['Cpu'].apply(lambda x:x.split()[0])
df['CPU_Company'].value_counts()


Intel      1240
AMD          62
Samsung       1
Name: CPU_Company, dtype: int64

3. Вывести на экран первые 5 строк
1. 1 Создать новый признак Cpu_Company, который будет содержать только название фирмы, которая произвела CPU (Intel, AMD и т.д)
1. 3 Создать новый признак Memory_Type, который будет содержать только тип носителя (HDD, SDD и т.д.)
1. 4 Удалите признаки Memory и ScreenResolution


1. 2 Создать новый признак Memory_Amount, который будет содержать только количество Gb памяти без указания типа носителя (128, 1024, 256 и т.д.)

In [24]:
# df['Memory_Amount'] = 
df['Memory'].unique()

a ='1.0TB'
print(a[-2:])
a [:-2]
int(a[0])

TB


1

In [25]:
df['Memory'].value_counts()

256GB SSD                        412
1TB HDD                          223
500GB HDD                        132
512GB SSD                        118
128GB SSD +  1TB HDD              94
128GB SSD                         76
256GB SSD +  1TB HDD              73
32GB Flash Storage                38
2TB HDD                           16
64GB Flash Storage                15
512GB SSD +  1TB HDD              14
1TB SSD                           14
256GB SSD +  2TB HDD              10
1.0TB Hybrid                       9
256GB Flash Storage                8
16GB Flash Storage                 7
32GB SSD                           6
180GB SSD                          5
128GB Flash Storage                4
512GB SSD +  2TB HDD               3
16GB SSD                           3
512GB Flash Storage                2
1TB SSD +  1TB HDD                 2
256GB SSD +  500GB HDD             2
128GB SSD +  2TB HDD               2
256GB SSD +  256GB SSD             2
512GB SSD +  256GB SSD             1
5

Задание 3.		
Создайте датафрейм с клиентами:					
clients = pd.DataFrame({
'client_id': [45, 32, 67, 33, 43], 'laptop_id': [506, 398, 710, 120, 1999]		
})				
laptop_id - это индексы датафрейма с ноутбуками			
3.1 Присоедините к таблице clients данные по ноутбукам через метод join						
3.2 Присоедините к таблице clients данные по ноутбукам через метод merge			

In [28]:
clients = pd.DataFrame({'client_id': [45, 32, 67, 33, 43],
},index = [506, 398, 710, 120, 1999])
clients

client_id
506          45
398          32
710          67
120          33
1999         43

In [37]:
joined_table = clients.join(df,how='inner')

In [38]:
joined_table

client_id Company            Product  TypeName  Inches  \
506         45  Lenovo      ThinkPad L570  Notebook    15.6   
710         67  Lenovo  IdeaPad 320-15ABR  Notebook    15.6   
120         33    Dell      Inspiron 7570  Notebook    15.6   

                ScreenResolution                          Cpu  Ram  \
506  IPS Panel Full HD 1920x1080   Intel Core i5 7200U 2.5GHz  8GB   
710                     1366x768  AMD A12-Series 9720P 3.6GHz  8GB   
120  IPS Panel Full HD 1920x1080   Intel Core i7 8550U 1.8GHz  8GB   

                   Memory                    Gpu       OpSys  Weight  \
506             256GB SSD  Intel HD Graphics 620  Windows 10   2.3kg   
710               1TB HDD          AMD Radeon R7  Windows 10   2.2kg   
120  256GB SSD +  1TB HDD   Nvidia GeForce 940MX  Windows 10  2.16kg   

     Price_euros CPU_Company  
506       911.00       Intel  
710       429.00         AMD  
120      1130.33       Intel

In [67]:

df["merge_index"]=df.index
df2=df[['Company',"Product",'merge_index','TypeName','Price_euros']]
df2

Company                              Product  merge_index  \
1      Apple                          MacBook Pro            1   
2      Apple                          Macbook Air            2   
3         HP                               250 G6            3   
4      Apple                          MacBook Pro            4   
5      Apple                          MacBook Pro            5   
...      ...                                  ...          ...   
1316  Lenovo                       Yoga 500-14ISK         1316   
1317  Lenovo                       Yoga 900-13ISK         1317   
1318  Lenovo                   IdeaPad 100S-14IBR         1318   
1319      HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon         1319   
1320    Asus  X553SA-XX031T (N3050/4GB/500GB/W10)         1320   

                TypeName  Price_euros  
1              Ultrabook      1339.69  
2              Ultrabook       898.94  
3               Notebook       575.00  
4              Ultrabook      2537.45  
5              Ultrabook      1803.60  
...                  ...          ...  
1316  2 in 1 Convertible       638.00  
1317  2 in 1 Convertible      1499.00  
1318            Notebook       229.00  
1319            Notebook       764.00  
1320            Notebook       369.00  

[1303 rows x 5 columns]

In [72]:
merged_table_left = clients.merge(df2, left_index=True,right_on="merge_index",how='left')
# clients.index
merged_table_left

client_id Company            Product  merge_index  TypeName  \
506.0         45  Lenovo      ThinkPad L570          506  Notebook   
NaN           32     NaN                NaN          398       NaN   
710.0         67  Lenovo  IdeaPad 320-15ABR          710  Notebook   
120.0         33    Dell      Inspiron 7570          120  Notebook   
NaN           43     NaN                NaN         1999       NaN   

       Price_euros  
506.0       911.00  
NaN            NaN  
710.0       429.00  
120.0      1130.33  
NaN            NaN

In [73]:
merged_table_outer = clients.merge(df2, left_index=True,right_on="merge_index",how='outer')
merged_table_outer

client_id Company                              Product  merge_index  \
506.0        45.0  Lenovo                        ThinkPad L570          506   
NaN          32.0     NaN                                  NaN          398   
710.0        67.0  Lenovo                    IdeaPad 320-15ABR          710   
120.0        33.0    Dell                        Inspiron 7570          120   
NaN          43.0     NaN                                  NaN         1999   
...           ...     ...                                  ...          ...   
1316.0        NaN  Lenovo                       Yoga 500-14ISK         1316   
1317.0        NaN  Lenovo                       Yoga 900-13ISK         1317   
1318.0        NaN  Lenovo                   IdeaPad 100S-14IBR         1318   
1319.0        NaN      HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon         1319   
1320.0        NaN    Asus  X553SA-XX031T (N3050/4GB/500GB/W10)         1320   

                  TypeName  Price_euros  
506.0             Notebook       911.00  
NaN                    NaN          NaN  
710.0             Notebook       429.00  
120.0             Notebook      1130.33  
NaN                    NaN          NaN  
...                    ...          ...  
1316.0  2 in 1 Convertible       638.00  
1317.0  2 in 1 Convertible      1499.00  
1318.0            Notebook       229.00  
1319.0            Notebook       764.00  
1320.0            Notebook       369.00  

[1305 rows x 6 columns]

In [74]:
merged_table_right = clients.merge(df2, left_index=True,right_on="merge_index",how='right')
merged_table_right

client_id Company                              Product  merge_index  \
1           NaN   Apple                          MacBook Pro            1   
2           NaN   Apple                          Macbook Air            2   
3           NaN      HP                               250 G6            3   
4           NaN   Apple                          MacBook Pro            4   
5           NaN   Apple                          MacBook Pro            5   
...         ...     ...                                  ...          ...   
1316        NaN  Lenovo                       Yoga 500-14ISK         1316   
1317        NaN  Lenovo                       Yoga 900-13ISK         1317   
1318        NaN  Lenovo                   IdeaPad 100S-14IBR         1318   
1319        NaN      HP  15-AC110nv (i7-6500U/6GB/1TB/Radeon         1319   
1320        NaN    Asus  X553SA-XX031T (N3050/4GB/500GB/W10)         1320   

                TypeName  Price_euros  
1              Ultrabook      1339.69  
2              Ultrabook       898.94  
3               Notebook       575.00  
4              Ultrabook      2537.45  
5              Ultrabook      1803.60  
...                  ...          ...  
1316  2 in 1 Convertible       638.00  
1317  2 in 1 Convertible      1499.00  
1318            Notebook       229.00  
1319            Notebook       764.00  
1320            Notebook       369.00  

[1303 rows x 6 columns]

Company                                          Apple
Product                                    MacBook Pro
TypeName                                     Ultrabook
Inches                                            13.3
ScreenResolution    IPS Panel Retina Display 2560x1600
Cpu                               Intel Core i5 2.0GHz
Ram                                                8GB
Memory                                       256GB SSD
Gpu                            Intel Iris Graphics 540
OpSys                                            macOS
Weight                                          1.37kg
Price_euros                                     1419.0

In [52]:
df.iloc[506]

Company                                   Asus
Product                 ZenBook UX510UX-CN211T
TypeName                              Notebook
Inches                                    15.6
ScreenResolution             Full HD 1920x1080
Cpu                 Intel Core i7 7500U 2.7GHz
Ram                                        8GB
Memory                    256GB SSD +  1TB HDD
Gpu                      Intel HD Graphics 620
OpSys                               Windows 10
Weight                                     2kg
Price_euros                             1224.0
Name: 513, dtype: object